Fix duplicated UUID migration issue

blizzz opened the pull request:
solves #30936

strategy based on https://github.com/nextcloud/server/issues/30936#issuecomment-1027844910

* [x] invalidate duplicated UUIDs in migration step
* [x] make it possible to re-run migration step for broken setup
* [x] occ command to let Nextcloud try to update UUIDs to current value¹
* [x] add a warning in admin overview if invalidated UUIDs are present
* [x] testing

¹ could be a manually triggerable repair step, but it would run once, be less flexible and this way it MAY be also easier for some LDAP server migration scenarios when an update was needed anyways (but only works of DN stays identical).

blizzz committed:
invalidated duplicated UUIDs prior to migration change

- in a proper setup there are no duplicated UUIDs
- not all setups are proper
- log warning to admin

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
allow to re-run migration step when backup DB is already created

- to heal incomplete states from broken upgrades

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

come-nc commented on apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:
```diff
@@ -52,6 +53,23 @@ public function getName() {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+		if ($schema->hasTable('ldap_group_mapping_backup')) {
+			// Previous upgrades of a broken release might have left an incomplete
+			// ldap_group_mapping_backup table. No need to recreate, but it
+			// should be empty.
```
My fear here is that depending where the update failed, the backup table may be the only one still having the mapping information. Maybe add an extra check that ldap_group_mapping is existing and not empty (unless backup is empty as well)?

blizzz replied:
`ldap_group_mapping` can be legitimately empty when groups are not configured.

Where could this migration step currently fail? 

1. in `changeSchema`, then we would have no backup table at all and therefore this is irrelevant.
2. in `postSchemaChange`, but this affects updating ldap_dn_hashes in the "old" tables only, no deletion or copying.

Therefore, I think we are safe in this context.

blizzz replied:
I think I did not look into Version1120Date20210917155206 though (discarded if for the lower version number), I shall have a closer look.

blizzz replied:
Actually truncating is already over-defensive for it should be empty nevertheless. This table is created in this migration step, but not populated yet. The check introduced in `schemaChange` is sufficient, and the truncation block can be dropped entirely.

blizzz replied:
But… then this migration step should have worked already priorly in #30936 yet, so why would it be executed again? 🤔

come-nc commented on apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:
```diff
@@ -172,4 +190,86 @@ protected function getUpdateQuery(string $table): IQueryBuilder {
+		$idsWithUuidToInvalidate = [];
+		foreach ($uuids as $uuid) {
+			array_push($idsWithUuidToInvalidate, ...$this->getNextcloudIdsByUuid($table, $uuid));
+		}
```
Missing a call to invalidateUuids here I suppose?

blizzz replied:
Oh my, I was already a step ahead in my mind.

blizzz committed:
add occ command to update UUIDs (incomplete)

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

come-nc committed:
Fix ldap:update-uuid

Generators cannot be iterated with while or returned by an other
 generator, using foreach instead.
And a few other problems.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix psalm errors in apps/user_ldap/lib/Jobs/CleanUp.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

blizzz committed:
Setup warning for invalid LDAP user or group UUIDs.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Added the checkup warning and did some basic testing with  the occ command, which was working as expected.

Minor annoyance: we have no means to clean up (i.e. remove) old gone groups. That's not a blocker and can be worked around with custom  instructions if necessary.

nickvergessen committed:
Don't provide favorite activity settings

Since mails and notifications are only available for actions of other users
it does not make sense to allow changing this.
It also prevents the common misunderstanding with
"file was changed inside a favorited folder"

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Temporarily disable samba-non-native as it times out too often

Signed-off-by: Joas Schilling <coding@schilljs.com>

blizzz commented:
Rebased and squashes the fixups. My testing was successful so far.

PVince81 commented on apps/user_ldap/lib/Jobs/CleanUp.php:
```diff
@@ -119,38 +119,29 @@ public function setArguments($arguments) {
 		}
 		$users = $this->mapping->getList($this->getOffset(), $this->getChunkSize());
-		if (!is_array($users)) {
-			//something wrong? Let's start from the beginning next time and
```
is this obsolete in general or connected to the context of this PR ?

blizzz replied:
Obsolete for return type hinting.

PVince81 reviewed (commented):
I flew over the code and didn't find anything notable apart from my minor comments.

I hope others can chime in to review the more complex code more in detail,

blizzz commented on apps/user_ldap/lib/Command/UpdateUUID.php:
```diff
@@ -0,0 +1,374 @@
+use Symfony\Component\Console\Output\OutputInterface;
+use function sprintf;
+
+class UuidUpdateReport {
```
Legal it is. I believe it is also not the sole case in our code bases, but given, I would not know for certain where to look at first. It is an internal helper class that just contains public properties, in order to not use an arbitrary array with with arbitrary keys.

come-nc reviewed (approved):
:+1:

PVince81 reviewed (approved):
:+1:

there's a DCO error though

PVince81 commented:
/backport to stable23

PVince81 commented:
/backport to stable22

PVince81 commented:
/backport to stable21

PVince81 commented:
/backport to stable20

PVince81 commented:
down to 20 because the original fix was put there as well...

come-nc commented:
I messed up my `git rebase -i` and duplicated two commits from Joas.
Not sure how that happenned but it causes no trouble on master because the commits were already on master.
For instance https://github.com/nextcloud/server/commit/1ab0ed81346f77e4e9faab64f7e02977c2849e1e is a duplicate of https://github.com/nextcloud/server/commit/5f5142ff583dfdc9ae82b01177c839f164bce2dc

I will create the backport by hand for 23 and backport from there.

ChristophWurst commented:
Command is not documented in admin manual :crying_cat_face:
